service-base/src/main/java/org/eclipse/hono/service/http/HttpContextTenantAndAuthIdProvider.java

```diff
@@ -0,0 +1,112 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.opentracing.SpanContext;
+import io.vertx.core.Future;
+import io.vertx.core.http.HttpHeaders;
+import io.vertx.ext.web.RoutingContext;
+
+/**
+ * Provides a method to determine the tenant and auth-id of a HTTP request from the given HttpContext.
+ */
+public class HttpContextTenantAndAuthIdProvider extends BaseExecutionContextTenantAndAuthIdProvider
+        implements ExecutionContextTenantAndAuthIdProvider<HttpContext> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(HttpContextTenantAndAuthIdProvider.class);
+
+    private final String tenantIdContextParamName;
+    private final String deviceIdContextParamName;
+
+    /**
+     * Creates a new HttpContextTenantAndAuthIdProvider.
+     * 
+     * @param config The configuration.
+     * @param tenantClientFactory The factory to use for creating a Tenant service client.
+     * @param tenantIdContextParamName The name of the HttpContext parameter name that provides the tenant id in case of
+     *            an unauthenticated request.
+     * @param deviceIdContextParamName The name of the HttpContext parameter name that provides the device id in case of
+     *            an unauthenticated request. The device id will be used as auth-id for the created
+     *            TenantObjectWithAuthId objects.
+     */
+    public HttpContextTenantAndAuthIdProvider(final ProtocolAdapterProperties config,
+            final TenantClientFactory tenantClientFactory,
+            final String tenantIdContextParamName,
+            final String deviceIdContextParamName) {
+        super(config, tenantClientFactory);
+        this.tenantIdContextParamName = tenantIdContextParamName;
+        this.deviceIdContextParamName = deviceIdContextParamName;
+    }
+
+    @Override
+    public Future<TenantObjectWithAuthId> get(final HttpContext context, final SpanContext spanContext) {
+        if (config.isAuthenticationRequired()) {
+            return getTenantViaCert(context.getRoutingContext(), spanContext)
+                    .recover(thr -> getTenantFromAuthHeader(context.getRoutingContext(), spanContext));
+        }
+        final String tenantId = context.get(tenantIdContextParamName);
+        if (tenantId != null) {
+            // unauthenticated request
+            final String deviceId = context.get(deviceIdContextParamName);
+            return tenantClientFactory.getOrCreateTenantClient()
+                    .compose(tenantClient -> tenantClient.get(tenantId, spanContext))
+                    .map(tenantObject -> new TenantObjectWithAuthId(tenantObject, deviceId));
+        }
+        return Future.failedFuture("tenant could not be determined");
+    }
+
+    private Future<TenantObjectWithAuthId> getTenantViaCert(final RoutingContext ctx, final SpanContext spanContext) {
+        if (!ctx.request().isSSL()) {
+            return Future.failedFuture("no cert found (not SSL/TLS encrypted)");
```

Comment:
it is quite expensive to create the stack trace associated with the failed future. Since this outcome will be very common, e.g. if all devices use Basic auth, we should probably better return `null` as the value in a succeeded future in this case. Same holds true for Basic auth ...

Another option would be to (re-)use a statically initialized exception instance for creating the failed future ...

WDYT?

Reply:
`Future.failedFuture(String)` will create a `NoStackTraceThrowable` (where `Throwable.fillInStackTrace()` doesn't get invoked). Therefore I don't think using failed futures here is more expensive than using succeeded futures.

Reply:
Didn't know that `Future.failedFuture(String)` doesn't create a stack trace :+1: